File: 0221-maximal-square/0221-maximal-square.py

"""

max int 

if cell == 1


get the current square of the current position


from x,y 

find the width and the height

[1] => 1

[0] => 0



all the members of a square have at least 2 neighbors 




max height 
max width



"""

class Solution:
    def maximalSquare(self, matrix: List[List[str]]) -> int:
        coln, rown = len(matrix), len(matrix[0])
        dp = [[0] * (rown+1) for _ in range(coln+1)]
        m = 0 
        print(dp)
        for col in range(1,coln+1):
            for row in range(1, rown+1):
                if matrix[col-1][row-1] == "1":
                    dp[col][row] = min(dp[col][row-1], dp[col-1][row], dp[col-1][row-1]) +1
                    m = max(dp[col][row], m)       

        return m * m 

File: 0221-maximal-square/0221-maximal-square.go
func getMax(a,b int)int{
    if a > b {
        return a
    }
    return b
}

func getMin(a,b int)int{
    if a < b {
        return a
    }
    return b
}


func maximalSquare(matrix [][]byte) int {
    var max int
    dp := [][]int{}
    n := len(matrix)
    m := len(matrix[0])
    for i := 0 ; i<= n ; i++ {
        dp = append(dp, make([]int,m+1))
    }
    
    
    
    for x:= 1; x <= n ; x++ {
        for y := 1 ; y <= m; y++{
            top := dp[x-1][y]
            left := dp[x][y-1]
            corner := dp[x-1][y-1]
            if matrix[x-1][y-1] == '0'{
                continue
            }            
            
            dp[x][y] = getMin(top, getMin(left, corner)) + 1
            max = getMax(max, dp[x][y])
        }
    }
    
    
    return max * max
}